src/tiledb/cloud/utilities/profiler.py

```diff
@@ -0,0 +1,194 @@
+import subprocess
+import threading
+import time
+from typing import Optional
+
+import numpy as np
+
+import tiledb
+from tiledb.cloud.utilities import max_memory_usage
+
+
+def create_log_array(uri: str):
+    """
+    Create an array to hold log events.
+
+    :param uri: array URI
+    """
+
+    int_fl = tiledb.FilterList(
+        [
+            tiledb.DoubleDeltaFilter(),
+            tiledb.BitWidthReductionFilter(),
+            tiledb.ZstdFilter(),
+        ]
+    )
+    ascii_fl = tiledb.FilterList([tiledb.ZstdFilter()])
+
+    max_time = np.iinfo(np.uint64).max - 1
+    d0 = tiledb.Dim(
+        name="time_ms", dtype=np.uint64, domain=(0, max_time), filters=int_fl
+    )
+    dom = tiledb.Domain([d0])
+
+    a0 = tiledb.Attr(name="id", dtype="ascii", filters=ascii_fl)
+    a1 = tiledb.Attr(name="op", dtype="ascii", filters=ascii_fl)
+    a2 = tiledb.Attr(name="data", dtype="ascii", filters=ascii_fl)
+    a3 = tiledb.Attr(name="extra", dtype="ascii", filters=ascii_fl)
+
+    schema = tiledb.ArraySchema(
+        domain=dom,
+        sparse=True,
+        attrs=[a0, a1, a2, a3],
+        offsets_filters=int_fl,
+        allows_duplicates=True,
+    )
+
+    schema.check()
+    tiledb.Array.create(uri, schema)
+
+
+def write_log_event(
+    uri: str,
+    id: str,
+    op: Optional[str] = "",
+    data: Optional[str] = "",
+    extra: Optional[str] = "",
+):
+    """
+    Write an event to the log array.
+
+    :param uri: array URI
+    :param id: event id
+    :param op: event operation, defaults to ""
+    :param data: event data, defaults to ""
+    :param extra: event extra data, defaults to ""
+    """
+
+    t_now_ms = time.time() * 1000
+    with tiledb.open(uri, "w") as A:
+        A[t_now_ms] = {
+            "id": [id],
+            "op": [op],
+            "data": [data],
+            "extra": [extra],
+        }
+
+
+class Profiler(object):
```

Comment:
It seems like it might make more sense to have this as a context manager so you can do:

```python
with my_profiler.profile(whatever_param):
  ...
```

This might also help with the potential dangling array below (`self.array = tiledb.open(uri, "w")`)